File: src/apps/api/Controllers/LocationDataController.cs
using System.Net;
using Microsoft.AspNetCore.Mvc;
using Api.Repository;
using System.Text.Json;

namespace Api.Controllers;

[ApiController]
[Route("[controller]")]
public class LocationDataController : ControllerBase
{
    private readonly LocationDataRepository _locationDataRepository;

    public LocationDataController()
    {
        _locationDataRepository = new LocationDataRepository();
    }

    [HttpGet]
    [Route("chatbotapikey")]
    public async Task<IActionResult> GetChatbotApiKey()
    {
        try
        {
            var key = await _locationDataRepository.GetChatbotApiKey();
            return Ok(key);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    [HttpGet]
    [Route("mapboxkey")]
    public async Task<IActionResult> GetMapBoxKey()
    {
        try
        {
            var key = await _locationDataRepository.GetMapBoxApiKey();
            return Ok(key);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    [HttpGet]
    [Route("googlemapskey")]
    public async Task<IActionResult> GetGoogleMapsKey()
    {
        try
        {
            var key = await _locationDataRepository.GetGoogleMapsKey();
            return Ok(key);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }



    [HttpGet]
    [Route("GetLocationData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.GetLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(404, "Solar Irradiation not found");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }


    }

    // GetEssentialLocationData
    [HttpGet]
    [Route("EssentialData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetEssentialLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.EssentialLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get essential location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }


    }

    // GetInitialLocationData
    [HttpGet]
    [Route("InitialData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetInitialLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.InitialLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get initial location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

        // GetSatelliteImageLocationData
    [HttpGet]
    [Route("SatelliteImageData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetSatelliteImageLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.SatelliteImageLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get satellite image location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    // GetMaskLocationData
    [HttpGet]
    [Route("MaskData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetMaskLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.MaskLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get mask location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

        // GetAnnualFluxLocationData
    [HttpGet]
    [Route("AnnualFluxData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetAnnualFluxLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.AnnualFluxLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get annual flux location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    // GetMonthlyFluxLocationData
    [HttpGet]
    [Route("MonthlyFluxData/{latitude}/{longitude}")]
    public async Task<IActionResult> GetMonthlyFluxLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            LocationDataModel data = await _locationDataRepository.MonthlyFluxLocationData(latitude, longitude);
            if (data == null)
            {
                return StatusCode(400, "Could not get monthly flux location data");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    [HttpPost]
    [Route("create")]
    public async Task<IActionResult> CreateLocationData([FromBody] LocationDataModel locationData)
    {
        try
        {
            string data = await _locationDataRepository.CreateLocationData(
                locationData.latitude,
                locationData.longitude,
                locationData.locationName!,
                JsonSerializer.Serialize(locationData.solarPanelsData!),
                locationData.satteliteImageData!,
                locationData.annualFluxData!,
                locationData.monthlyFluxData!,
                locationData.maskData!,
                locationData.horisonElevationData!
            );
            if (data.Equals("LocationData created successfully"))
            {
                return Ok(data);
            }
            return StatusCode(400, data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }



    [HttpDelete]
    [Route("delete/{latitude}/{longitude}")]
    public async Task<IActionResult> DeleteLocationData(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            string data = await _locationDataRepository.DeleteLocationData(latitude, longitude);
            if (data.Equals("LocationData not found"))
            {
                return StatusCode(404, "Solar Irradiation not found");
            }
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }

    // checkIfLocationDataExists
    [HttpGet]
    [Route("checkIfLocationDataExists/{latitude}/{longitude}")]
    public async Task<IActionResult> CheckIfLocationDataExists(
        [FromRoute] double latitude,
        [FromRoute] double longitude
    )
    {
        try
        {
            bool data = await _locationDataRepository.CheckIfLocationDataExists(latitude, longitude);
            return Ok(data);
        }
        catch (Exception e)
        {
            return StatusCode(500, e.Message);
        }
    }
}

File: src/apps/services/src/controllers/location.data/location.data.controller.ts

import { Request, Response } from 'express';
import { connection as conn } from '../../main';
import * as tedious from 'tedious';
import ILocationData from '../../models/location.data.interface';
export default class LocationDataController {
  public getChatBotApiKey = async (req: Request, res: Response) => {
    try {
      const key: string = process.env.CHATBOT_API_KEY;
      res.status(200).json(key);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };
  public getMapBoxApiKey = async (req: Request, res: Response) => {
    try {
      const key: string = process.env.MAP_BOX_API_KEY;
      res.status(200).json(key);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  public getGoogleApiKey = async (req: Request, res: Response) => {
    try {
      const key: string = process.env.GOOGLE_API_KEY;
      res.status(200).json(key);
    } catch (error) {
      res.status(500).json({ error: error });
    }
  };

  public createLocationData = async (req: Request, res: Response) => {
    const {
      latitude,
      longitude,
      locationName,
      solarPanelsData,
      satteliteImageData,
      annualFluxData,
      monthlyFluxData,
      maskData,
      horisonElevationData,
    } = req.body;
    const dateCreated = new Date().toISOString().slice(0, 19).replace('T', ' ');
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));
    const query = `INSERT INTO [dbo].[locationData] (latitude, longitude, locationName, solarPanelsData, satteliteImageData, annualFluxData, monthlyFluxData, maskData, dateCreated, horisonElevationData) VALUES (${lat}, ${long}, '${locationName}', '${solarPanelsData}', '${satteliteImageData}', '${annualFluxData}', '${monthlyFluxData}', '${maskData}', '${dateCreated}', '${horisonElevationData}')`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log(err.message);
            res.status(400).json({
              error: err.message,
            });
          } else {
            console.log(rowCount);
            res.status(200).json({
              message: 'Solar Irradiation created successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  public getLocationData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));
    const query = `SELECT latitude, longitude, locationName, solarPanelsData, satteliteImageData, annualFluxData, monthlyFluxData, maskData, dateCreated, horisonElevationData  FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;
    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            console.log('express error: ' + err.message);
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: columns[2].value,
          solarPanelsData: columns[3].value,
          satteliteImageData: columns[4].value,
          annualFluxData: columns[5].value,
          monthlyFluxData: columns[6].value,
          maskData: columns[7].value,
          dateCreated: columns[8].value,
          horisonElevationData: columns[9].value,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getEssentialLocationData
  public getEssentialData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, locationName, satteliteImageData, dateCreated FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: columns[2].value,
          satteliteImageData: columns[3].value,
          dateCreated: columns[4].value,
          annualFluxData: null,
          monthlyFluxData: null,
          maskData: null,
          solarPanelsData: null,
          horisonElevationData: null,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getInitialLocationData
  public getInitialData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, locationName, solarPanelsData, dateCreated, horisonElevationData FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: columns[2].value,
          satteliteImageData: null,
          dateCreated: columns[4].value,
          annualFluxData: null,
          monthlyFluxData: null,
          maskData: null,
          solarPanelsData: columns[3].value,
          horisonElevationData: columns[5].value,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getSatelliteImageLocationData
  public getSatelliteImageData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, satteliteImageData FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: null,
          satteliteImageData: columns[2].value,
          dateCreated: null,
          annualFluxData: null,
          monthlyFluxData: null,
          maskData: null,
          solarPanelsData: null,
          horisonElevationData: null,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getMaskLocationData
  public getMaskData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, maskData FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: null,
          satteliteImageData: null,
          dateCreated: null,
          annualFluxData: null,
          monthlyFluxData: null,
          maskData: columns[2].value,
          solarPanelsData: null,
          horisonElevationData: null,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getAnnualFluxLocationData
  public getAnnualFluxData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, annualFluxData FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: null,
          satteliteImageData: null,
          dateCreated: null,
          annualFluxData: columns[2].value,
          monthlyFluxData: null,
          maskData: null,
          solarPanelsData: null,
          horisonElevationData: null,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // getMonthlyFluxLocationData
  public getMonthlyFluxData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));

    const query = `SELECT latitude, longitude, monthlyFluxData FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    let solarIrradiation: ILocationData;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json(solarIrradiation);
          }
        }
      );

      request.on('row', (columns) => {
        solarIrradiation = {
          latitude: columns[0].value,
          longitude: columns[1].value,
          locationName: null,
          satteliteImageData: null,
          dateCreated: null,
          annualFluxData: null,
          monthlyFluxData: columns[2].value,
          maskData: null,
          solarPanelsData: null,
          horisonElevationData: null,
        };
      });

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  //delete deleteLocationData
  public deleteLocationData = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));
    const query = `DELETE FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;

    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            console.log(rowCount);
            res.status(200).json({
              message: 'Solar Irradiation deleted successfully.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };

  // check if location data exists
  public checkIfLocationDataExists = async (req: Request, res: Response) => {
    const { latitude, longitude } = req.params;
    const lat = parseFloat(latitude.replace(',', '.'));
    const long = parseFloat(longitude.replace(',', '.'));
    const query = `SELECT latitude, longitude, locationName FROM [dbo].[locationData] WHERE latitude = ${lat} AND longitude = ${long}`;
    console.log('Checking if location data exists ' + lat + ', ' + long);
    try {
      const request = new tedious.Request(
        query,
        (err: tedious.RequestError, rowCount: number) => {
          if (err) {
            res.status(400).json({
              error: err.message,
            });
          } else if (rowCount === 0) {
            res.status(404).json({
              message: 'Solar Irradiation not found.',
            });
          } else {
            res.status(200).json({
              message: 'Solar Irradiation found.',
            });
          }
        }
      );

      conn.execSql(request);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  };
}
